cunumeric/eager.py

```diff
@@ -171,25 +189,31 @@ class EagerArray(NumPyThunk):
     :meta private:
     """
 
-    def __init__(self, runtime, array, parent=None, key=None) -> None:
+    def __init__(
+        self,
+        runtime: Runtime,
+        array: Any,
+        parent: Optional[Any] = None,
+        key: Optional[Any] = None,
+    ) -> None:
         super().__init__(runtime, array.dtype)
-        self.array = array
-        self.parent = parent
-        self.children = None
+        self.array: Any = array
+        self.parent: Any = parent
+        self.children: list[Any] = []
         self.key = key
         #: if this ever becomes set (to a DeferredArray), we forward all
         #: operations to it
-        self.deferred = None
+        self.deferred: Optional[DeferredArray] = None
         self.escaped = False
 
     @property
-    def storage(self):
+    def storage(self) -> Any:
         if self.deferred is None:
             self.to_deferred_array()
-        return self.deferred.storage
+        return self.deferred.storage  # type: ignore
```

Comment:
mypy does not know `self.deferred` is no longer None after the call with side effects. If `self.to_deferred_array()` could instead return a value that is assigned, that would make the type flow clearer